pangeo_forge_recipes/recipes/reference_hdf_zarr.py

```diff
@@ -0,0 +1,205 @@
+
+    Output parameters:
+
+    :param output_url: where the final reference file is written.
+    :param output_storage_options: dict of kwargs for creating fsspec
+        instance when writing final output
+    :param template_count: the number of occurrences of a URL before it
+        gets made a template. ``None`` to disable templating
+    """
+
+    # TODO: support chunked ("tree") aggregation: would entail processing each file
+    #  in one stage, running a set of merges in a second step and doing
+    #  a master merge in finalise. This would maybe map to iter_chunk,
+    #  store_chunk, finalize_target. The strategy was used for NWM's 370k files.
+
+    # TODO: as written, this rcipe is specific to HDF5 files. fsspec-reference-maker
+    #  also supports TIFF and grib2 (and more coming)
+    netcdf_url: Union[str, List[str]]
+    output_url: str
+    output_target: FSSpecTarget
+    _work_dir: MetadataTarget
+    netcdf_storage_options: dict = field(default_factory=dict)
+    inline_threshold: int = 500
+    output_storage_options: dict = field(default_factory=dict)
+    template_count: Optional[int] = 20
+    xarray_open_kwargs: dict = field(default_factory=dict)
+    xarray_concat_args: dict = field(default_factory=dict)
+
+    def iter_inputs(self) -> Iterable[Hashable]:
+        return ()
+
+    @property
+    def cache_input(self) -> Callable[[Hashable], None]:
+        return no_op
+
+    @property
+    def prepare_target(self) -> Callable[[], None]:
+        """Prepare the recipe for execution by initializing the target.
+        Attribute that returns a callable function.
+        """
+        return no_op
+
+    def iter_chunks(self) -> Iterable[Hashable]:
+        """Iterate over all target chunks."""
+        return fsspec.open_files(self.netcdf_url, **self.netcdf_storage_options)
+
+    @property
+    def store_chunk(self) -> Callable[[Hashable], None]:
+        """Store a chunk of data in the target.
+        """
+        return functools.partial(_one_chunk, work_dir=self._work_dir)
+
+    @property
+    def finalize_target(self) -> Callable[[], None]:
+        """Final step to finish the recipe after data has been written.
+        Attribute that returns a callable function.
+        """
+        if isinstance(self.netcdf_url, str):
+            proto = fsspec.utils.get_protocol(self.netcdf_url)
+        else:
+            proto = fsspec.utils.get_protocol(self.netcdf_url[0])
+        return functools.partial(
+            _finalize,
+            work_dir=self._work_dir,
+            out_url=self.output_url,
+            out_target=self.output_target,
+            out_so=self.output_storage_options,
+            remote_protocol=proto,
+            remote_options=self.netcdf_storage_options,
+            xr_open_kwargs=self.xarray_open_kwargs,
+            xr_concat_kwargs=self.xarray_concat_args,
+            template_count=self.template_count,
+        )
+
+
+def _one_chunk(of, work_dir):
```

Comment:
Following up on my earlier comment, to align with XarrayZarrRecipe, the idea here would be
- accept the `chunk_key` as an argument (instead of of), plus the FilePattern
- look up the filename using `file_pattern[chunk_key]`
- open the file using `storage.file_opener`
- proceed with the rest of the recipe